src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,104 @@
 package core.basesyntax;
 
+import java.util.Arrays;
+import java.util.NoSuchElementException;
+
 public class ArrayList<T> implements List<T> {
+    private static final int INITIAL_CAPACITY = 10;
+    private static final double GROW_RATE = 1.5;
+    private int size;
+    private T[] elementData;
+
+    public ArrayList() {
+        elementData = (T[]) new Object[INITIAL_CAPACITY];
+    }
+
     @Override
     public void add(T value) {
-
+        ensureCapacity();
+        elementData[size] = value;
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
-
+        validateAddIndex(index);
+        ensureCapacity();
+        System.arraycopy(elementData, index, elementData, index + 1, size - index);
+        elementData[index] = value;
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        for (int i = 0; i < list.size(); i++) {
+            add(list.get(i));
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        validateIndex(index);
+        return elementData[index];
     }
 
     @Override
     public void set(T value, int index) {
+        validateIndex(index);
+        elementData[index] = value;
 
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        validateIndex(index);
+        T removedElement = elementData[index];
+        System.arraycopy(elementData, index + 1, elementData, index, size - index - 1);
+        elementData[--size] = null;
+        return removedElement;
     }
 
     @Override
     public T remove(T element) {
-        return null;
+        for (int i = 0; i < size; i++) {
+            if (elementData[i] != null && elementData[i].equals(element)
+                    || elementData[i] == element) {
+                return remove(i);
+            }
+        }
+        throw new NoSuchElementException("Element not found in list");
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private void grow() {
+        int newCapacity = (int)(elementData.length * GROW_RATE);
```

Comment:
```suggestion
        int newCapacity = (int) (elementData.length * GROW_RATE);
```
make sure to autoformat your code in IDEA